tidy3d/components/parameter_perturbation.py

```diff
@@ -0,0 +1,722 @@
+    def _sample(
+        self, temperature: Union[ArrayLike[float], SpatialDataArray]
+    ) -> Union[ArrayLike[float], ArrayLike[Complex], SpatialDataArray]:
+        """New sample function."""
+
+        # if temperature range not present just return original function
+        if self.temperature_range is None:
+            return sample(self, temperature)
+
+        temh_min, temh_max = self.temperature_range
+        if np.any(temperature < temh_min) or np.any(temperature > temh_max):
+            log.warning(
+                "temperature passed to 'HeatPerturbation.sample()'"
+                f"is outside of 'HeatPerturbation.temperature_range' = {self.temperature_range}"
+            )
+        return sample(self, temperature)
+
+    return _sample
+
+
+class HeatPerturbation(AbstractPerturbation):
+    """Abstract class for heat perturbation."""
+
+    temperature_range: Tuple[pd.NonNegativeFloat, pd.NonNegativeFloat] = pd.Field(
+        (0, inf),
+        title="Temperature range",
+        description="Temparature range in which perturbation model is valid.",
+        units=KELVIN,
+    )
+
+    @abstractmethod
+    def sample(
+        self, temperature: Union[ArrayLike[float], SpatialDataArray]
+    ) -> Union[ArrayLike[float], ArrayLike[Complex], SpatialDataArray]:
+        """Sample perturbation.
+
+        Parameters
+        ----------
+        temperature : Union[ArrayLike[float], SpatialDataArray]
+            Temperature sample point(s).
+
+        Returns
+        -------
+        Union[ArrayLike[float], ArrayLike[Complex], SpatialDataArray]
+            Sampled perturbation value(s).
+        """
+
+    @add_ax_if_none
+    def plot(
+        self,
+        temperature: ArrayLike[float],
+        val: FieldVal = "real",
+        ax: Ax = None,
+    ) -> Ax:
+        """Plot perturbation using provided temperature sample points.
+
+        Parameters
+        ----------
+        temperature : ArrayLike[float]
+            Array of temperature sample points.
+        val : Literal['real', 'imag', 'abs', 'abs^2', 'phase'] = 'real'
+            Which part of the field to plot.
+        ax : matplotlib.axes._subplots.Axes = None
+            Matplotlib axes to plot on, if not specified, one is created.
+
+        Returns
+        -------
+        matplotlib.axes._subplots.Axes
+            The supplied or created matplotlib axes.
+        """
+
+        temperature_numpy = np.array(temperature)
+
+        values = self.sample(temperature_numpy)
+        values = self._get_val(values, val)
+
+        ax.plot(temperature_numpy, values)
+        ax.set_xlabel("temperature (K)")
+        ax.set_ylabel(f"{val}(perturbation value)")
+        ax.set_title("temperature dependence")
+        ax.set_aspect("auto")
+
+        return ax
+
+
+class LinearHeatPerturbation(HeatPerturbation):
+    """Linear heat perturbation."""
+
+    temperature_ref: pd.NonNegativeFloat = pd.Field(
+        ...,
+        title="Reference temperature",
```

Comment:
I think some of the docstrings and descriptions could use a bit more expanding on. To someone not familiar with this, it could be hard to understand what these classes and their fields do. For example, the docstring "Linear heat perturbation" could be "Specification describing how heat quantities affect optical medium properties." and the description of "reference temperature" could be something like "temperature at which the thermo optic coefficient is valid. Away from this temperature ..."

Your notebook has a lot of really useful descriptions, maybe some can be copied into the docs too.